WIP: Overview Documentation for PSF Fitting Workflow (#766)
astropy/photutils: docs/psf_spec/high_level_overview.rst

to implement specific versions of each part of the fitting process, implementing
a variety of fitting procedures best suited to the specific science cases. However,
at its most basic, all choices share the same basic functionality: a 2D image is
supplied and a `~astropy.table.Table` of detected sources, with positions and flux
counts, is returned.

This simplified workflow is expanded below for the case of the
`~photutils.psf.IterativelySubtractedPSFPhotometry` class, building upon the simpler
`~photutils.psf.BasicPSFPhotometry` algorithm. While there are additional blocks
used within the workflow, the main `~photutils.psf.BasicPSFPhotometry` flow is
``image`` :raw-html:`&rarr;` ``finder`` :raw-html:`&rarr;` ``group maker``
:raw-html:`&rarr;` ``fitter`` :raw-html:`&rarr;` ``stars``. The extension to more
complex fitting classes, such as `~photutils.psf.IterativelySubtractedPSFPhotometry`,
have the flow ``image`` :raw-html:`&rarr;` ``finder`` :raw-html:`&rarr;` ``group maker``
:raw-html:`&rarr;` ``fitter`` :raw-html:`&rarr;` ``culler and ender`` :raw-html:`&rarr;`
``finder`` :raw-html:`&rarr;` ... :raw-html:`&rarr;` ``stars``.

> Each submodule is described briefly here, with context for how it fits into the overall

Reviewer: Minor fix to Erik's suggestion: A brief summary of each block is described here...

> Each submodule is described briefly here, with context for how it fits into the overall

Reviewer: Minor fix to Erik's suggestion: A brief summary of each block is described here...

PSF fitting framework, providing a top-down view of the process. Where available, links
to more detailed descriptions of each block is also provided.

> .. graphviz::

Reviewer 2: In general this is great!  One thing about this, though: It might be helpful to have some lines that pass through the image.  That is, have lines from where the image "enters" the big box to the finder and the background estimator?

And another thing: maybe try to add a "legend" that shows what the lines mean?  E.g. "solid = required flow of data", "dashed = optional information flow" and "dotted = optional component" or something like that?

Reviewer: I'll second this. I was also going to suggest a description of what the different lines mean.

    digraph PSFFitting {
        compound = true;
        newrank = true;
        subgraph cluster0 {
            label="IterativelySubtractedPSFPhotometry"
            f [label = "finder"];
            g [label = "group maker"];
            som [label = "single object model"];
            n [label = "noise"];
            ft [label = "fitter"];
            b [label = "background estimator"];
            c [label = "culler and ender"];
            // required to allow finder and guess to combine into group maker
            invis [shape = circle, width = .1, fixedsize = true, style = invis];
        }

        w [label = "wcs"];
        i [label = "input image", shape = box];
        gs [label = "guess at stars", shape = box];
        p [label = "psf model"];
        sm [label = "scene maker"];

>         st [label = "stars\n(astropy.table)", shape = box];

Reviewer 2: I wonder of "stars" here should be "objects" or something like that?  Presumably the scene maker at least *might* have the option to do other kinds of fitting?

Reviewer: I second this also

        labelloc = top;

        w -> i [style = dashed];
        // lhead sets the arrow end at the subgraph, rather than at finder itself
        i -> f [lhead = cluster0];
        gs -> invis [style = dashed];

        f -> g;
        f -> invis;
        invis -> g;
        som -> f [style = dashed];
        n -> ft;
        som -> ft;
        g -> ft;
        b -> f [style = dashed];
        b -> g [style = dashed];
        b -> ft [style = dashed];
        ft -> c [style = dashed, label = "Subtract", dir = none];
        c -> f [style = dashed];

        i -> gs [style = invis];
        
        p -> som;
        sm -> g [style = dotted];
        ft -> st;
        // reverse arrow to put psf model above scene maker, saving space with scene
        // maker next to the middle of the main box
        p -> sm [style = dotted, dir = back];

        // puts things that should be at the top of the box at the top for orientation
        // and structuring; requires 'newrank' above to set subgraph items equal
        {rank = same; w; som;}
        {rank = same; p; f; n;}
        {rank = same; gs; g; sm;}
    }

.. _Image:

Input Image
^^^^^^^^^^^

> The input image is the most fundamental part of the PSF fitting process, as it is the

Reviewer 2: Should probably say something here about `NDData` objects.  That is, the fundamental object we will want to use to pass around the images is probably `NDData`.  There can just be something then that says roughly "but there's a way in that could instead use raw arrays or FITS files that then get auto-converted to `NDData"?  Or that might be a detail that can instead be in the other documents.

Reviewer: Since there are details on the content of the image here, I agree with Erik that NDData objects should be mentioned.

product on which all subsequent methodology is applied. The image must be a two-dimensional
array of ``x`` and ``y`` coordinates with ``flux`` represented by array values. If a
WCS -- contained within a FITS file, for instance -- is passed with the two-dimensional
array then the position of sources can be given in sky coordinates, instead of pixel values.
The given image is passed as an input to the chosen PSF fitting class.

.. _Finder:

Finder
^^^^^^

> The finder is the first step in the PSF fitting process, as sources must be discovered in

Reviewer 2: These all need to be implementation of the `StarFinderBase` class, right? (i.e., https://photutils.readthedocs.io/en/stable/api/photutils.StarFinderBase.html#photutils.StarFinderBase).  If so you should say that, and link to it (e.g. ``  `~photutils.detectionStarFinderBase`  ``)

Reviewer 2: More broadly: I wonder if the `StarFinderBase` should actually be `ObjectFinderBase` or something like that?  Not necessary for this document, but worth thinking a bit about.

Author: I've updated the above references of ``stars`` to ``sources`` as that doesn't particularly affect anything and can be future-proofed now, but I obviously can't change these names as they're hard-coded links for now. Where would be best to raise this as a separate issue? Is actually making an issue the right approach for this?

the image before any kind of fit can be applied to them. These accept the input ``image``
and produce a `~astropy.table.Table` of detected sources, by a set of criteria internal to
the given ``Finder`` (see, e.g., `~photutils.detection.DAOStarFinder` using ``roundness``
and ``sharpness`` to determine if sources are point-like). If an initial set of detected
sources is passed to the fitter as ``init_guesses`` then ``Finder`` is not run on the first
pass of an iterative fitting class, instead using the provided positions and fluxes. See
the :doc:`finder <finder>` documentation for more details.

.. _Group Maker:

Group Maker
^^^^^^^^^^^

The second block to run in the fitting process, the various group maker processes, such as

> `~photutils.psf.DAOGroup`, allow for the merging of astrometrically overlapping sources

Reviewer 2: ```suggestion
 `~photutils.psf.DAOGroup`, allow for the merging of sources that are astrometrically nearby
```

"overlapping" is a bit unclear because the definition of "overlapping" somewhat depends on the algorithm.  I think "nearby" is better in that it's intentionally vague?

into a set of multiple stars. These stars must be fit simultaneously (see Fitter_) as a

> composite model, avoiding the potential issue where blended sources return incorrect fluxes,

Reviewer 2: ```suggestion
 composite model, making deblending of sources possible.
```
(contingent on the next line suggestion)

> as affects aperture photometry. The block accepts the `~astropy.table.Table` output from

Reviewer 2: ```suggestion
The block accepts the `~astropy.table.Table` output from
```

Finder_ and returns a second `~astropy.table.Table` containing the same columns as the input
as well as an additional column indicating the group number of each source. More information
on this block can be found in :doc:`Group Maker <group_maker>`.

.. _Fitter:

Fitter
^^^^^^
Once a set of sources has been detected and grouped, each source must have its respective

> properties -- position and flux primarily -- determined. For this a ``Fitter`` instance is

Reviewer 2: ```suggestion
 properties -- always position and flux, but sometimes other properties -- determined. For this a ``Fitter`` instance is
```

required; these are drawn from a separate class or instance of minimization routines, such as

> those given under the `~astropy.modeling` umbrella. The fitter must be a two-dimensional fitter,

Reviewer 2: ```suggestion
 those implemented in `~astropy.modeling.fitting`. The fitter must be a two-dimensional fitter,
```

by nature of the image to be fit, accepted as an input to the instance. The class must then

> return the position of the source (or sources, if a composite model was supplied; see

Reviewer 2: I think this part on the exact inputs/outputs could be left out because it's actually a bit more complicated (i.e., it has to conform to the `astropy.modeling.fitting` fitter API).  So just saying "it has to take in the image and table and output the fitted model.  More detail given in ..." or something like that is sufficient.

`Group Maker`_) and its total flux, as determined by the PSF model describing the telescope
system -- optics, CCD, etc. -- that created the image. See `below <PSF Model_>`_ for more
details on the PSF model, and the :doc:`fitter <fitter>` documentation for further
information on the ``Fitter`` instance and its properties.

.. _PSF Model:

PSF Model
^^^^^^^^^

> Perhaps the most important aspect of PSF fitting, the ``PSF Model`` describes the distribution

Reviewer 2: ```suggestion
 Perhaps the most variable aspect of PSF fitting, the ``PSF Model`` describes the distribution
```
?

Author: I was trying to lead into this section with the idea that while the rest of the sections (getting good sources, grouping blended things, etc.) are important, *this* section has to be done right or there's no point doing any of the other stuff at all... So important is what I meant (instead of variable, as in "most likely to change for each user"?), but this is slightly qualitative and perhaps subjective language; none of the other sections have lead-in.

I think the question is more:
do we just start this section with ```The ``PSF model`` describes the distribution```?

of light falling on a given pixel from a source at some position with some total flux, fully
describing the effects of telescope optics, quantized CCD pixels, etc. In this respect it is

> more properly described as a PRF than a PSF -- see the :doc:`documentation <psf_model>` for

Reviewer 2: I think it would be good to also link to the "terminology" section here: https://photutils.readthedocs.io/en/latest/psf.html#terminology (that is, link using sphinx syntax to the point in the docs that that link goes to), since that discusses the PRF/PSF/etc thing.

Reviewer: agreed. Or at the very least, define the acronyms.

> more details -- but we conflate the two terms here freely. Similar to the ``Fitter``, it is

Reviewer 2: ```suggestion
 more details -- but we use a single term here for brevity. Similar to the ``Fitter``, it is
```
we should probably fix this at some point and use the terminology more consistently?  But that can be done later, probably.

Author: I've made the change to the language, but to the more overall point: I'm not sure we really *do* need to change the terminology. Literally nowhere in the `photutils` framework do we NOT do PRF fitting, by nature of fitting quantised images we can't do anything but. Thus so long as we make it clear that technically PSFs can be (and are) pedantically different and that we are actually doing PRF fitting but we're just gonna use the colloquialism if that's ok with everyone else I think we get away with it.

Instead of putting loads of work into rewriting all of our documentation to use the new and more proper language, we just have to put work into making sure at any necessary point we differentiate between the two and link to the terminology page (like above!).

a separate callable class detailing a specific PSF response. This can be some analytical
function, such as `~photutils.psf.IntegratedGaussianPRF`, or an empirical description of the
PSF, such as that implemented by the effective PSF functionality (see :ref:`build-epsf` for
details). The model must describe a two-dimensional distribution, and must return the total
flux of the source and its sub-pixel position via the ``evaluate`` function, accepted by the

> Fitter_ when minimizing the fit between the model and the data.

Reviewer 2: As above with the fitter, I think it's best to describe it as it has to follow the `astropy.modeling.models` API for 2D models, because there's a lot of "little things" there that may end up mattering a lot.

.. _noise:

Noise Description
^^^^^^^^^^^^^^^^^
The ``noise`` block within the PSF fitting process is, in much the same way as the ``Fitter``
block, a separate ``callable`` function mapping the relationship between the pixel values in

> the input ``image`` and the corresponding uncertainty -- in units of standard deviations -- of

Reviewer 2: I'm not sure I like the "in units of standard deviation part", but will discuss more in #755

Author: I've removed the subclause for now as it's not particularly necessary (this being a brief, high level overview and all) but we can discuss that further in the other PR and I'll make any changes finalised there in this branch too.

the pixel. It must therefore accept a two-dimensional array and return an array of the same
shape as ``image``. Alternatively it can be overloaded to indicate that the uncertainty array
was passed along with ``image`` via an `~astropy.nddata.NDData` instance. A more detailed
description of this block is available :doc:`here <noise_data>`.

.. _Background Estimator:

Background Estimator
^^^^^^^^^^^^^^^^^^^^
The ``image`` on which source extraction and evaluation is to be performed is not necessarily
assumed to have had any pre-processing applied to it, and thus it may be necessary to account
for the counts of the image background when handling PSF fitting. This block handles

> the determination of the background levels, avoiding high count peaks to calculate the typical

Reviewer 2: what do you mean by "avoiding high count peaks"?  I think that might be algorithm-specific, so maybe better to just do this:
```suggestion
 the determination of the background levels, and should provide the typical
```
?

count of an otherwise empty pixel. The block must therefore accepts the two-dimensional ``image``
and returns either a single value -- the background count of the entire image, as calculated
by the criteria of the specific block implementation -- or a 2D array of values, allowing for
the possibility of varying background counts across the image. More information on this block
can be found :doc:`here <background_estimator>`.

.. _Culler and Ender:

Culler and Ender
^^^^^^^^^^^^^^^^

> Only implemented in iterative fitting classes, part of the extension to

Reviewer 2: I don't understand the "part of the extension to" part... partly because `BasicPSFPhotometry` isn't really defined here (see my general comment)

`~photutils.psf.BasicPSFPhotometry`, this block is the final step of a single iteration. While
a maximum number of iterations can be specified for, e.g., 
`~photutils.psf.IterativelySubtractedPSFPhotometry`, this block -- specifically, the ender half
-- assesses whether all sources have been found within the ``image``. If no new sources have
been found within an iteration, fitting can be stopped prematurely, avoiding wasteful computational
time. Similarly, the culler aspect of this block examines sources found within an iteration for
quality; it should calculate some goodness-of-fit criterion and reject sources picked up by the
``Finder`` but fall below a given threshold. This block therefore accepts the ``stars``
output `~astropy.table.Table` and returns a new `~astropy.table.Table` with low quality sources
-- such as cosmic ray hits present in the image -- removed as well as a boolean flag indicating
whether the iterative fitting process has reached a conclusion and can be terminated. Please refer
to the :doc:`documentation <culler_and_ender>` for further details.

.. _Single Object Model:

> Single Object Model

Reviewer 2: This (along with the scene maker) are *optional*, right?  That is, if not present, it means just use the PSF model directly?  In that case you should say that in both descriptions.

Author: I've added language to this effect in the text, but eventually I don't think they'll be entirely optional, in that I think the block will always have to be used, but it can have convenient "do nothing" defaults. I think it's tied to the ``object_type`` flag that will eventually have to be thrown around in the ``astropy.table.Table`` objects: if that (by default) is something like ``point source`` then ``Single Object Model`` simply returns the input PSF model (i.e., f * &#948; = f) and ``Scene Maker`` just returns the input table of objects (i.e., no input sources are an extended source split apart).

This *effectively* means they're optional, as the user can just ignore that the more complex functionality exists and default parameters unchanged give the same answer as if they weren't there, but I think that's easier than having ``if use_som: new_model = single_object_model(psf)`` or equivalent.

^^^^^^^^^^^^^^^^^^^
An additional block sitting between the `PSF model`_ and the Fitter_, the ``Single Object Model``
extends the fitting capability of the class from purely point-source objects to extended sources.
This block, effectively, convolves the PSF of the given observation with a physical light
distribution, producing the flux seen in each pixel by the system of such an extended source. This
block therefore requires as its inputs the PSF model and the object type -- a string representing
the type of source (point-like, spiral galaxy, etc.) to create the intrinsic light distribution
for -- and returns a new `~astropy.modeling.Fittable2DModel` instance, the convolution of the PSF

> and the true source. Further specifics on this block, including how it will link to `Scene Maker`_

Reviewer 2: ```suggestion
 and the true source. Further specifics on this block, including how it connects to `Scene Maker`_
```

can be found in the :doc:`documentation <single_object_model>`.

.. _Scene Maker:

Scene Maker
^^^^^^^^^^^
The final block within the scope of PSF fitting, the scene maker is also the last to be implemented.
Currently no PSF fitting class extends its functionality to include this block, but it will
eventually generalize and possibly replace the `Group Maker`_ as the method of merging and assigning
non-independent detections within a given image. While current grouping is simply the determining
of point sources with potentially overlapping flux, the *scene* maker will allow for the grouping
of both point sources into an overlapping, simultaneous fit group but also into a single extended
object. Thus each iteration will require a step in which multiple-point-source extended objects are
evaluated for their separation, and newly detected point sources are evaluated for potential

> assignment as part of an extended source. The details of this block are available :doc:`here

Reviewer 2: This may be better discussed in #756 ... but this discussion makes me realize that maybe the scene maker and single object model should be one thing?  Or even if separate classes, maybe they should be described together in one document?

Author: They're basically "point + extended" sources extensions of group maker and the PSF model, so while linked slightly more obviously by the need to share a common ``object_type`` list and know what to do with each available choice in each instance, they're no more or less inter-connected than group maker and PSF model; it's just obvious what to do with those as the only potential object type is a point source, which is simple to handle.

We can discuss this further in the corresponding PR to get into the details, and, again, I'll make any changes decided upon in there in this branch too.
